packages/loader/container-loader/src/deltaManager.ts

```diff
@@ -606,11 +610,23 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
         // Drop pending messages - this will ensure catchUp() does not go into infinite loop
         this.pending = [];
 
+        if (emitDisposed === true) {
+            this.disposeInternal(error);
+        } else {
+            this.emit("closed", error);
+        }
+    }
+
+    private disposeInternal(error?: ICriticalContainerError): void {
+        if (this._disposed) {
+            return;
+        }
+        this._disposed = true;
+
         // This needs to be the last thing we do (before removing listeners), as it causes
         // Container to dispose context and break ability of data stores / runtime to "hear"
         // from delta manager, including notification (above) about readonly state.
-        this.emit("closed", error);
-
+        this.emit("disposed", error);
         this.removeAllListeners();
```

Comment:
I get why you can't do this for `close` anymore, but it makes me a little nervous in terms of causing some memory leak in so-called back compat cases, where the host calls close only.  What do you think?

Off the cuff, one idea would be to use a direct callback instead of events to notify container of DM disposal, then you can leave this in place on close.

Reply:
What if, in this PR in main, at the end of DM.close (after emitting close event) it immediately calls dispose?  You might even be able to make the overall PR closer to the next one, with that one addition.

Reply:
I think this might overcomplicate the code to ensure the new usage of close and dispose also works as expected (in terms of when we are expecting the "closed" and "disposed" events from Container/DM)